### ISSUE TITLE
Padding in AES-256 CBC mode: multiple of key size or multiple of block size?

### ISSUE THREAD

**Issue by MatthiasValvekens** [wbond/oscrypto#52]
I'm in the process of implementing PDF 2.0 encryption, which is AES-256 based. At some point, the algorithm requires encrypting a specific 16-byte string using AES-256, without padding against an all-zero IV (so it's effectively one block of unpadded AES-256 in ECB mode).

This is the relevant snippet:

```
        perms_bytes = b'\xfc\xff\xff\xff'
        extd_perms_bytes = (
            perms_bytes + (b'\xff' * 4) + b'Tadb' + secrets.token_bytes(4)
        )
        perms = struct.unpack('<i', perms_bytes)[0]
        _, encrypted_perms = symmetric.aes_cbc_no_padding_encrypt(
            encryption_key, extd_perms_bytes, bytes(16)
        )
```

My tests pass on my macOS machine, but the OpenSSL-based implementation used on Linux throws an error here.

```
2021-01-14T23:32:31.2930994Z         if cipher != 'rc4' and not padding:
2021-01-14T23:32:31.2931561Z             # AES in CBC mode can be allowed with no padding if
2021-01-14T23:32:31.2932148Z             # the data is an exact multiple of the key size
2021-01-14T23:32:31.2932655Z             aes128_no_padding = (
2021-01-14T23:32:31.2933212Z                 cipher == 'aes128' and
2021-01-14T23:32:31.2933653Z                 padding is False and
2021-01-14T23:32:31.2934052Z                 len(data) % 16 == 0
2021-01-14T23:32:31.2934382Z             )
2021-01-14T23:32:31.2934762Z             aes192_no_padding = (
2021-01-14T23:32:31.2935390Z                 cipher == 'aes192' and
2021-01-14T23:32:31.2935779Z                 padding is False and
2021-01-14T23:32:31.2936133Z                 len(data) % 24 == 0
2021-01-14T23:32:31.2936436Z             )
2021-01-14T23:32:31.2936747Z             aes256_no_padding = (
2021-01-14T23:32:31.2937334Z                 cipher == 'aes256' and
2021-01-14T23:32:31.2937728Z                 padding is False and
2021-01-14T23:32:31.2938079Z                 len(data) % 32 == 0
2021-01-14T23:32:31.2938379Z             )
2021-01-14T23:32:31.2938890Z             if aes128_no_padding is False and aes192_no_padding is False and aes256_no_padding is False:
2021-01-14T23:32:31.2939704Z >               raise ValueError('padding must be specified')
2021-01-14T23:32:31.2940252Z E               ValueError: padding must be specified
```

So correct me if I'm wrong, but shouldn't the correct requirement be to force the data length to be a multiple of the _block_ size (i.e. 16 bytes), and not the _key_ size (32 bytes)? I'm happy to submit a pull request rectifying that, but I wanted to get a second opinion first.

EDIT: thanks a lot for your efforts, by the way, I've been using your crypto suite for quite some time now, it's really been a tremendous asset!

**Comment by MatthiasValvekens** [comment]
I was able to reproduce the issue by explicitly importing the OpenSSL implementation of `aes_cbc_no_padding_encrypt`. Modifying the padding test in `_encrypt` to test modulo 16 instead of 32 fixes the problem. That said, it appears that `aes_cbc_no_padding_encrypt` already tests whether the input data length is divisible by 16, so the check in `_encrypt` might not even be necessary at all.

What do you think? :)

**Comment by wbond** [comment]
Sorry for the delay in responding to this.

The check was malformed for the AES ciphers, but not in a way that resulting in invalid/bad encryption. Instead, it just prevented certain valid values from being encrypted.

This should be fixed as of 7a8906149d6ed898f2622c627dee342e9e87a34f, and will be part of the 1.3.0 release.
